### ISSUE TITLE
Roslyn assemblies are failed to be loaded by test runners

### ISSUE THREAD

**Issue by xperiandri** [dotnet/roslyn#4886]
I tried to run unit tests within Visual Studio and it cannot discover test for all the unit test projects except `CSharpVisualStudioTest` project and sometimes two another.

See output from test runners by link:
https://github.com/xunit/xunit/issues/562#issuecomment-136018490

What is the problem?

**Comment by Pilchie** [comment]
The problem is that the VS test runner shadow copies the test assemblies, and the CLR won't load "Fake Signed" binaries when shadow copying.  I believe @KevinH-MS took a look at this before.

**Comment by xperiandri** [comment]
Does this mean that it will not ever possible to run Roslyn unit tests from Test Explorer?

**Comment by KevinH-MS** [comment]
@xperiandri - that is correct.  Test Explorer will not load the tests unless strong name signing is disabled.

**Comment by xperiandri** [comment]
Does this mean that if I will myself disable signing in project properties it will work?
Do I need to disable it for all projects?

Where can I read about what does fake signing means?

**Comment by KevinH-MS** [comment]
Sorry for the rude delay in responding here...  Yes, for now, the easiest thing to do is disable signing for all projects.  The easiest way to do this would be to set an environment variable (SignAssembly=false) and then start devenv.  Alternatively, you could modify your ```build\Targets\VSL.Settings.targets``` to set SignAssembly=false.  ***Note:***  This may break other scenarios (like deploying your bits to an experimental instance of VS).

As for FakeSign, there's no documentation, but the tool is here:  https://github.com/dotnet/roslyn/tree/master/src/Tools/Source/FakeSign

Basically, it's just setting the "signed" bit in metadata (even though the assembly isn't fully signed).  For most scenarios, the CLR will then treat the assembly as being signed (without requiring you to add a global "skip verification" entry, which requires admin privileges).

The newest release of xUnit allows you to control AppDomain behavior on a more granular level, and I'm hopeful that we might be able to utilize that to get things working.  I'm using #210 to track this.

**Comment by xperiandri** [comment]
Thanks for explanation especially about fake signing.
It is nice to know that some work is going on.
Now I temporary use CodeRush to run Unto Tests.

**Comment by xperiandri** [comment]
@KevinH-MS, it is not that simple as just disabling FakeSign.
I changed line 89 `    <SignAssembly Condition="'$(SignAssembly)' == ''">true</SignAssembly>` from true to false as you suggested but solution does not build because InternalsVisibleTo requires strong name.
So I got tens of thousands of errors :)

**Comment by KevinH-MS** [comment]
What command/project are you using to build?  ```msbuild /m Roslyn.sln``` should work after this modification (the target that generates the InternalsVisibleTo should respect the SignAssembly property).

**Comment by xperiandri** [comment]
I tried .\cibuild.cmd
And build within VS

**Comment by KevinH-MS** [comment]
Ah...I see.  Try setting this as well:
```
build/Targets/VSL.Settings.targets:

-    <UseRoslynAnalyzers Condition="'$(UseRoslynAnalyzers)' == ''">true</UseRoslynAnalyzers>
+    <UseRoslynAnalyzers Condition="'$(UseRoslynAnalyzers)' == ''">false</UseRoslynAnalyzers>
```

**Comment by KevinH-MS** [comment]
Also note:  If the solution builds fine in VS, but the errors still show up in the Error List, you might have to delete the .vs directory (next to Roslyn.sln).  There's a bug in the project system that may cause it to not update.  It will be fixed in the next Update to VS 2014 (sometime this Fall).

**Comment by davkean** [comment]
Right now - the Test Explorer inside Visual Studio cannot scale to Roslyn's ~60,000 tests - it runs out of memory. While the VS owners are looking into making it scale - there's no work for us to do here, we're not going to enable them to show up until this is fixed.
